src/txrequest.h

```diff
@@ -0,0 +1,200 @@
+    const std::unique_ptr<Impl> m_impl;
+
+public:
+    //! Construct a TxRequestTracker.
+    TxRequestTracker(bool deterministic = false);
+    ~TxRequestTracker();
+
+    // Conceptually, the data structure consists of a collection of entries, one for each peer/txhash combination
+    // (an "announcement"):
+    //
+    // - CANDIDATE entries represent transactions that were announced by a peer, and that become available for
+    //   download after their reqtime has passed.
+    //
+    // - REQUESTED entries represent transactions that have been requested, and which we're awaiting a response for
+    //   from that peer. Their expiry value determines when the request times out.
+    //
+    // - COMPLETED entries represent transactions that have been requested from a peer, and a NOTFOUND or a
+    //   transaction was received in response (valid or not), or they timed out. They're only kept around to
+    //   prevent requesting them again. If only COMPLETED entries for a given txhash remain (so no CANDIDATE or
+    //   REQUESTED ones), all of them are deleted (this is an invariant, and maintained by all operations below).
+    //
+    // The operations below manipulate the data structure.
+
+    /** Deletes all entries for a given peer.
+     *
+     * It should be called when a peer goes offline.
+     */
+    void DisconnectedPeer(uint64_t peer);
+
+    /** Deletes all entries for a given txhash (both txid and wtxid ones).
+     *
+     * This should be called when a transaction is successfully added to the mempool, seen in a block, or for
+     * whatever reason we no longer care about it.
+     */
+    void ForgetTxHash(const uint256& txhash);
+
+    /** Adds a new CANDIDATE announcement.
+     *
+     * Does nothing if one already exists for that (txhash, peer) combination (whether it's CANDIDATE, REQUESTED, or
+     * COMPLETED). Note that the txid/wtxid property is ignored for determining uniqueness, so if an announcement
+     * is added for a wtxid H, while one for txid H from the same peer already exists, it will be ignored. This is
+     * harmless as the txhashes being equal implies it is a non-segwit transaction, so it doesn't matter how it is
+     * fetched. The new announcement is given the specified preferred and reqtime values, and takes it is_wtxid from
+     * the specified gtxid.
+     */
+    void ReceivedInv(uint64_t peer, const GenTxid& gtxid, bool preferred,
+        std::chrono::microseconds reqtime);
+
+    /** Converts the CANDIDATE announcement for the provided peer and gtxid into a REQUESTED one.
+     *
+     * Expiry is set to the specified value. This can ONLY be called immediately after GetRequestable was called
+     * (for the same peer), with only ForgetTx and other RequestedTx calls (both for other txhashes) in
+     * between. Any other non-const operation removes the ability to call RequestedTx.
+     */
+    void RequestedTx(uint64_t peer, const GenTxid& gtxid, std::chrono::microseconds expiry);
```

Comment:
Constructing vectors instead of repeatedly calling the function seems worse to me, fwiw, but agree that it would be nice to have a simpler description of the constraint.

Maybe tying it to the return value of `GetRequestable` might be better: "each value returned by `GetRequestable` may used with either `ForgetTxHash` or `RequestedTx` for the given peer, but not both, and at most only once. Any other calls to non-const methods on `TxRequestTracker` invalidate all the results from `GetRequestable`" ?

Reply:
It wouldn't be too hard to just drop the constraint, and instead specify it as:
* If no announcement with the specified (peer, txhash) exists, or it isn't in CANDIDATE state, the call has no effect.
* The specified announcement is changed from CANDIDATE to REQUESTED.
* If another announcement for the same txhash was already in REQUESTED state, it is marked COMPLETED.

It'd be a bit more code, but would make the function fully specified. I've avoided it, as there is really no point for that functionality in practice, but it does make things a bit nicer. WDYT?

If we don't do that, I like @ajtowns's way of describing the functionality.

Reply:
Either is fine. I also like @ajtowns's description.

Whichever way we go, I still think the function should be changed to take a txhash - I should have left that as a separate comment.

Reply:
Changed to AJ's description.

Reply:
Here is the patch to make `RequestedTx` fully specified:

```patch
diff --git a/src/txrequest.cpp b/src/txrequest.cpp
index af4b59755b..581b498180 100644
--- a/src/txrequest.cpp
+++ b/src/txrequest.cpp
@@ -590,11 +590,29 @@ public:
     void RequestedTx(NodeId peer, const uint256& txhash, std::chrono::microseconds expiry)
     {
         auto it = m_index.get<ByPeer>().find(ByPeerView{peer, true, txhash});
-        // RequestedTx can only be called on CANDIDATE_BEST announcements (this is implied by its condition that it
-        // can only be called on GenTxids returned by GetRequestable (and only ForgetTxHash and RequestedTx can be
-        // called in between, which preserve the state of other GenTxids).
-        assert(it != m_index.get<ByPeer>().end());
-        assert(it->GetState() == State::CANDIDATE_BEST);
+        if (it == m_index.get<ByPeer>().end()) {
+            // There is no CANDIDATE_BEST entry, look for _READY or _DELAYED instead.
+            it = m_index.get<ByPeer>().find(ByPeerView{peer, false, txhash});
+            if (it == m_index.get<ByPeer>().end() || (it->GetState() != State::CANDIDATE_DELAYED &&
+                it->GetState() != State::CANDIDATE_READY)) {
+                // There is no CANDIDATE_* entry at all, give up.
+                return;
+            }
+
+            // Look for an existing CANDIDATE_BEST to demote to _READY, or REQUESTED to make COMPLETED.
+            auto it_old = m_index.get<ByTxHash>().lower_bound(ByTxHashView{txhash, State::CANDIDATE_BEST, 0});
+            if (it_old != m_index.get<ByTxHash>().end() && it_old->m_txhash == txhash) {
+                if (it_old->GetState() == State::CANDIDATE_BEST) {
+                    // It doesn't matter whether we pick CANDIDATE_READY or _DELAYED here, as SetTimePoint()
+                    // will correct it at GetRequestable() time. If time only goes forward, it will always be
+                    // _READY, so pick that to avoid extra work in SetTimePoint().
+                    Modify<ByTxHash>(it_old, [](Announcement& ann) { ann.SetState(State::CANDIDATE_READY); });
+                } else if (it_old->GetState() == State::REQUESTED) {
+                    Modify<ByTxHash>(it_old, [](Announcement& ann) { ann.SetState(State::COMPLETED); });
+                }
+            }
+        }
+
         Modify<ByPeer>(it, [expiry](Announcement& ann) {
             ann.SetState(State::REQUESTED);
             ann.m_time = expiry;
```

When combined with the corresponding changes in src/test/fuzz/txrequest.cpp it's actually a net reduction in code. WDYT?

Reply:
Added as an extra commit on top.

Reply:
I think that without this change all the remaining asserts are redundant internal checks, rather than "invalid use of API" ones, which I think is a plus. Also makes perfect sense (in retrospect) that that would then make the fuzz tester a bunch simpler.

I think that means that you can call TxRequestTracker methods with any params and the behaviours well-defined, reasonably logical, and as efficient as possible. I think this even maintains the amortized complexity, which surprises me.

So yeah, I like this! Maybe change the "There is no CANDIDATE_BEST entry" comment to note that this path only occurs if the API caller requests an txid that wasn't just returned from `GetRequestable` -- ie, they either have a bug due to a race condition of not dealing with the results of GetRequestable for one peer before calling it again later, or they're doing their own selection of which announcement to request beyond relying on TxRequestTracker's algorithm?

Maybe the "give up"  case would be clearer if there was a bit more of the "why" spelled out too?

```c++
            if (it == m_index.get<ByPeer>().end()) {
                // This txid wasn't tracked for this peer; caller should have called ReceivedInv() first.
                return;
            } else if (it->GetState() != State::CANDIDATE_DELAYED && it->GetState() != State::CANDIDATE_READY)) {
                // Only progress to REQUESTED from CANDIDATE to ensure the state machine is finite
                return;
           }
```

Reply:
Done, switched to the fully-defined version (squashed into the appropriate commits). I added some of @ajtowns's comments, and more.

Reply:
I also like this. Good change!